packages/react/src/components/TreeView/TreeNode.js

```diff
@@ -0,0 +1,215 @@
+    [`${prefix}--tree-node--disabled`]: disabled,
+    [`${prefix}--tree-node--selected`]: selected.includes(value),
+    [`${prefix}--tree-node--with-icon`]: Icon,
+    [`${prefix}--tree-leaf-node`]: !children,
+    [`${prefix}--tree-parent-node`]: children,
+  });
+  const toggleClasses = classNames(`${prefix}--tree-parent-node__toggle-icon`, {
+    [`${prefix}--tree-parent-node__toggle-icon--expanded`]: expanded,
+  });
+  const handleToggleClick = event => {
+    if (onToggle) {
+      onToggle(event, { isExpanded: !expanded });
+    }
+    setExpanded(!expanded);
+  };
+  const handleClick = event => {
+    event.stopPropagation();
+    if (onSelect && !disabled) {
+      onSelect(event, { value });
+    }
+  };
+  const handleKeyDown = event => {
+    if (matches(event, [keys.ArrowLeft, keys.ArrowRight, keys.Enter])) {
+      event.stopPropagation();
+    }
+    if (children && match(event, keys.ArrowLeft)) {
+      setExpanded(false);
+    }
+    if (children && match(event, keys.ArrowRight)) {
+      setExpanded(true);
+    }
+    if (match(event, keys.Enter) && onSelect && !disabled) {
+      onSelect(event, { value });
+    }
+    if (rest.onKeyDown) {
+      rest.onKeyDown(event);
+    }
+  };
+
+  useEffect(() => {
+    /**
+     * Negative margin shifts node to align with the left side boundary of the
+     * tree
+     * Dynamically calculate padding to recreate tree node indentation
+     * - parent nodes have (depth + 1rem) left padding
+     * - leaf nodes have (depth + 2.5rem) left padding without icons (because
+     *   of expando icon + spacing)
+     * - leaf nodes have (depth + 2rem) left padding with icons (because of
+     *   reduced spacing between the expando icon and the node icon + label)
+     */
+    const calcOffset = () => {
+      // parent node
+      if (children) {
+        return depth + 1;
+      }
+      // leaf node with icon
+      if (Icon) {
+        return depth + 2;
+      }
+      // leaf node without icon
+      return depth + 2.5;
+    };
+    const currentNodeLabel = currentNode.current.querySelector(
+      `.${prefix}--tree-node__label`
+    );
+    currentNodeLabel.style.marginLeft = `-${calcOffset()}rem`;
+    currentNodeLabel.style.paddingLeft = `${calcOffset()}rem`;
+
+    // sync props and state
+    setExpanded(isExpanded);
+  }, [children, depth, Icon, isExpanded]);
+
+  const treeNodeProps = {
+    ...rest,
+    className: treeNodeClasses,
+    onClick: handleClick,
+    onKeyDown: handleKeyDown,
+    ref: currentNode,
+    role: 'treeitem',
+    tabIndex: (!disabled && -1) || null,
+  };
+
+  if (!children) {
+    return (
+      <li {...treeNodeProps}>
+        <div className={`${prefix}--tree-node__label`}>
+          {Icon && <Icon className={`${prefix}--tree-node__icon`} />}
+          {label}
+        </div>
+      </li>
+    );
+  }
+  return (
+    <li {...treeNodeProps} aria-expanded={isExpanded}>
+      <div className={`${prefix}--tree-node__label`}>
+        <button
```

Comment:
When I tried to keep hitting tab keys to get into tree view and kept hitting tab keys, I saw expandos got keyboard focuses. I believe it was not intentional?

```suggestion
        <button
          tabIndex={-1}
```